### PR TITLE
New: Add generic-type-naming (#111)

### PR DESCRIPTION
**Pull request by unlight**
https://github.com/nzakas/eslint-plugin-typescript/issues/111

### CODE REVIEW

File: docs/rules/generic-type-naming.md [hunk 1]
```diff
@@ -40,3 +40,5 @@ If you do not want to enforce a naming convention for type variables.
 
 ## Further Reading
 - [TypeScript Generics](https://www.typescriptlang.org/docs/handbook/generics.html)
+
+^[A-Z]+$
```
Comment (j-f1) [inline docs/rules/generic-type-naming.md]: Is this here on purpose?

Comment (unlight) [reply docs/rules/generic-type-naming.md]: No, by accident

File: lib/rules/generic-type-naming.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    params.forEach(node => {
+        const type = node.type;
+
+        if (type === "TSTypeParameter" || type === "TypeParameter") {
+            const name = node.name;
+
+            if (name && !regex.test(name)) {
+                const data = { name, rule };
+
+                context.report({ node, messageId: "nameRule", data });
```
Comment (macklinu) [inline lib/rules/generic-type-naming.js]: The `messageId` API was released in [ESLint v4.15.0](https://eslint.org/blog/2018/01/eslint-v4.15.0-released#new-apis). Is it backwards-compatible? I'm wondering if we should use the older API for compatibility with all versions of ESLint:

```js
context.report({ node, message, data })
```

File: lib/rules/generic-type-naming.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        const rule = context.options[0];
+
+        if (!rule) {
+            return {};
+        }
+        const regex = new RegExp(rule);
+
+        return {
+            VariableDeclarator: node =>
+                checkTypeParameters(node.typeParameters, context, regex, rule),
```
Comment (macklinu) [inline lib/rules/generic-type-naming.js]: I wonder if you could use a higher-order function to DRY up this section:

```js
function createTypeParameterChecker(context, rule) {
  const regex = new RegExp(rule)
  return function checkTypeParameters(node) {
    const params = node.typeParameters && node.typeParameters.params
    // ... rest of logic you already have
  }
}
```

And then your rule would look something like this:

```js
create(context) {
  const rule = context.options[0]
  if (!rule) {
    return
  }

  const checkTypeParameters = createTypeParameterChecker(context, rule)
  return {
    VariableDeclarator: checkTypeParameters, // `node` gets passed to this function
    ClassDeclaration: checkTypeParameters,
    InterfaceDeclaration: checkTypeParameters,
    // etc.
  }
}
```

I like higher-order functions for situations like this, but I wonder if anyone else has feedback on this / if it makes sense. 😄

File: docs/rules/generic-type-naming.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+interface SimpleMap<TValue> {
+  [key: string]: TValue;
+}
+```
+
+Examples of **incorrect** code with a configuration of `'^T[A-Z][a-zA-Z]+$'`:
+
+```typescript
+type ReadOnly<T extends object> = {
+  readonly [Key in keyof T]: TType[Key];
```
Comment (macklinu) [inline docs/rules/generic-type-naming.md]: Should `TType` read `T` in this line?

**Comment by JamesHenry** [comment]
@unlight I think @macklinu has provided some good feedback here. I will give you chance to comment and/or update before merging this.

**Comment by unlight** [comment]
I agree, feedback is good. I'll try to improve code in the coming days.

**Comment by JamesHenry** [comment]
@unlight let me know if you plan on making any other changes, or if this is now ready to go

**Comment by unlight** [comment]
@macklinu Please, take a look. I've implemented your code review notes.
@JamesHenry I'm not planning to make any other changes

**Comment by JamesHenry** [comment]
@unlight @macklinu I am very sorry for the long delay in coming back to this, we are going to work out a better maintenance plan for this project

**Pull request merged**